distributed/scheduler.py

```diff
@@ -3643,6 +3643,7 @@ def __init__(
             "dump_cluster_state_to_url": self.dump_cluster_state_to_url,
             "benchmark_hardware": self.benchmark_hardware,
             "get_story": self.get_story,
+            "check_idle": self.check_idle,
```

Comment:
@jacobtomlinson just curious, where is this RPC used? I don't see a corresponding client method, or use on workers. Are you planning to call it in dask-k8s?

Reply:
Yeah using it in dask-kubernetes, but only as a fallback if the HTTP API is not enabled.

Reply:
Good point, I actually missed this. I think "best practice" should be to register this in dask-k8s with a scheduler extension. If I just had a look at this code base, it would look like dead code and I might remove it.

We're rarely doing these refactorings but there is nothing here that tells us what is actually "public" and what isn't so there is always a risk

Reply:
The fallback of the fallback is to register a scheduler extension so that we can support older versions of distributed. My worry is that scheduler extensions still depend on scheduler internals like `Scheduler.idle_since` which could also change.

There are tests in dask-kubernetes that cover this, so in terms of risk there is some mitigation. But it would be really nice to try and move towards a world where we have well defined public APIs in terms of the scheduler object (for extensions), the RPC and the HTTP API. Otherwise it is always going to be risky developing against the scheduler.

We could also expose this via the client and treat that as the public API instead of prodding the RPC directly.

Reply:
> There are tests in dask-kubernetes that cover this, so in terms of risk there is some mitigation. But it would be really nice to try and move towards a world where we have well defined public APIs in terms of the scheduler object (for extensions), the RPC and the HTTP API. Otherwise it is always going to be risky developing against the scheduler.

This is a much bigger thing that I want it to be but yes I would love to see _nothing_ of the actual Scheduler object to be public. It is just too big, there are too many small things, it's changing too quickly, ...

> We could also expose this via the client and treat that as the public API instead of prodding the RPC directly.

That's typically something I feel very comfortable with since it is very explicit that this is intended for external usage.


---

Just to be very clear: It's fine keeping it like this. "Fixing" this API leakage is very hard. I'm open to any suggestions but handing users (both in extension but especially in plugins) the entire object isn't maintainable, particularly not as long as there is no proper discipline around "using underscores"

Reply:
> This is a much bigger thing that I want it to be but yes I would love to see nothing of the actual Scheduler object to be public.

I totally sympathise with that. Although scheduler plugins/extensions/run_on_scheduler are all well established patterns. So backing out from those at this point will be hard.

> That's typically something I feel very comfortable with since it is very explicit that this is intended for external usage.

Most, if not all, of the cluster managers open an RPC to the scheduler and invoke methods directly. I think part of the challenge here is that some cluster managers live in distributed and some live in `dask/dask{kubernetes,jobqueue,yarn,gateway,cloudprovider,etc}`. 

For a long time the `dask-foo` projects were considered part of core Dask and it wasn't unusual to make a PR to both distributed and the cluster manager you were working on which were coupled. However this is a maintenance challenge because tests for code in `distributed` run in other repos.

Perhaps a good step forward would be for `dask-foo` projects to only interact with the scheduler via the `Client` and the HTTP API. These are API surfaces that we can consider to be public. 

The question is what do we do with the base `Cluster` and `SpecCluster` classes in `distributed`? These both use the RPC directly and are intended to be subclassed by third-party libraries. Maybe these classes should be updated to stop using the RPC and to use a `Client` instead?